README.md
![2017071915004751124509.png](http://7xihe6.com1.z0.glb.clouddn.com/2017071915004751124509.png)

# What is ErGo?
An intelligent bot based on TensorFlow.

ErGo comes with the following choices:

- chat: simple chat,only support Englist now.


## Features 

* NLP
* RNN
* CNN
* Seq2seq
* QA

## Installation
The program requires the following dependencies (easy to install using pip):

- python 3.5+
- tensorflow (tested with v1.0+)
- numpy
- tqdm
- Flask (web interface)
- Redis (web interface)

## Usage

To train the model,simple run `main.py`.Before you chat ,you should train,just run `python3 main.py [-t chat] --train`.
Once trained,you can test the results with  `main.py [-t chat] --no-train` (have fun).

#### Results

```text
hello
-------------
master: hello
ergo:  hey .
``
What color ?
-------------
master: what color
ergo: i don 't know .
```

#### Data

- [data](https://dl.ysicing.net/download/tensorflow/)  
- [nltk](http://www.nltk.org/data.html) `nltk.download() `




ChangeLog
2017-07-19 Ysicing Zheng <root@ysicing.net>

    * Simple Demo

inference/chat.py
#!/usr/bin/env python3
# coding=utf-8
# Version:0.1
# Author: ysicing
# file: chat.py
# time: 2017/7/19 21:59
# Copyright 2017 ysicing. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------

import tensorflow as tf
import numpy as np
import utils.s2s_reader as s2s_reader
import sys
import os
import re

# -----------------------------------parameters------------------------------------------


# the file where we read the data/model from
file_name = "dataset"
# interactive mode allow user to talk to the model directly, if set to false, it will test on the training data instead
iteracitve = True
# regular expression for parsing user input
expression = r"[0-9]+|[']*[\w]+"
# signal mode allow user to insert signal token before the decoder generate sentence
signal = False
# batch size for testing
batch_size = 1

# data params
# bucket_option = [i for i in xrange(1, 20+1)]
bucket_option = [5, 10, 15, 20, 25, 31]
buckets = s2s_reader.create_bucket(bucket_option)

reader = s2s_reader.reader(file_name=file_name, batch_size=batch_size, buckets=buckets, bucket_option=bucket_option,
                           signal=signal)
vocab_size = len(reader.dict)

# if load_model = true, then we need to define the same parameter in the saved_model inorder to load it
hidden_size = 512
projection_size = 300
embedding_size = 300
num_layers = 1

# ouput_size for softmax layer
output_size = hidden_size
if projection_size != None:
    output_size = projection_size

# model name & save path
model_name = "p" + str(projection_size) + "_h" + str(hidden_size) + "_x" + str(num_layers)
save_path = file_name + "/" + model_name

# prediction params
beam_size = 10
top_k = 10
max_sequence_len = 20

# ---------------------------------model definition------------------------------------------

tf.reset_default_graph()
sess = tf.InteractiveSession()

# placeholder
enc_inputs = tf.placeholder(tf.int32, shape=(None, batch_size), name="enc_inputs")
targets = tf.placeholder(tf.int32, shape=(None, batch_size), name="targets")
dec_inputs = tf.placeholder(tf.int32, shape=(None, batch_size), name="dec_inputs")

# input embedding layers
emb_weights = tf.Variable(tf.random_normal([vocab_size, embedding_size]), name="emb_weights")
enc_inputs_emb = tf.nn.embedding_lookup(emb_weights, enc_inputs, name="enc_inputs_emb")
dec_inputs_emb = tf.nn.embedding_lookup(emb_weights, dec_inputs, name="dec_inputs_emb")


# cell definiton
def getStackedLSTM():
    cell_list = []
    for i in range(num_layers):
        single_cell = tf.contrib.rnn.LSTMCell(
            num_units=hidden_size,
            num_proj=projection_size,
            # initializer=tf.truncated_normal_initializer(stddev=truncated_std),
            state_is_tuple=True
        )
        cell_list.append(single_cell)
    return tf.contrib.rnn.MultiRNNCell(cells=cell_list, state_is_tuple=True)


# encoder & decoder defintion
_, enc_states = tf.nn.dynamic_rnn(cell=getStackedLSTM(),
                                  inputs=enc_inputs_emb,
                                  dtype=tf.float32,
                                  time_major=True,
                                  scope="encoder")

dec_outputs, dec_states = tf.nn.dynamic_rnn(cell=getStackedLSTM(),
                                            inputs=dec_inputs_emb,
                                            initial_state=enc_states,
                                            dtype=tf.float32,
                                            time_major=True,
                                            scope="decoder")

# output layers
project_w = tf.Variable(tf.truncated_normal([output_size, embedding_size], stddev=0.1), name="project_w")
project_b = tf.Variable(tf.constant(0.1, shape=[embedding_size]), name="project_b")
softmax_w = tf.Variable(tf.truncated_normal([embedding_size, vocab_size], stddev=0.1), name="softmax_w")
softmax_b = tf.Variable(tf.constant(0.1, shape=[vocab_size]), name="softmax_b")

dec_outputs = tf.reshape(dec_outputs, [-1, output_size])
dec_proj = tf.matmul(dec_outputs, project_w) + project_b
logits = tf.nn.log_softmax(tf.matmul(dec_proj, softmax_w) + softmax_b)

# prediction
logit = logits[-1]
top_values, top_indexs = tf.nn.top_k(logit, k=beam_size, sorted=True)

# load variable
saver = tf.train.Saver()
cwd = os.getcwd()
saver.restore(sess, cwd + "/" + save_path + "/model.ckpt")
print("\nModel restored.")


# ----------------------------prediciton helper function-----------------------------

def build_input(sequence):
    dec_inp = np.zeros((1, len(sequence)))
    dec_inp[0][:] = sequence
    return dec_inp.T


def print_sentence(index_list):
    for index in index_list:
        sys.stdout.write(reader.id_dict[index])
        sys.stdout.write(' ')
    sys.stdout.write('\n')


def predict(enc_inp):
    dec_inp = np.zeros((1, 1))
    dec_inp[0][0] = 2
    index_output = []
    feed_dict = {enc_inputs: enc_inp, dec_inputs: dec_inp}
    indexs, state = sess.run([top_indexs, dec_states], feed_dict)
    index_output.append(indexs[0])

    while True:
        dec_inp[0][0] = indexs[0]
        feed_dict = {enc_states: state, dec_inputs: dec_inp}
        indexs, state = sess.run([top_indexs, dec_states], feed_dict)
        if indexs[0] == 3:
            break
        index_output.append(indexs[0])

    return index_output


def beam_predict(enc_inp, signal=None):
    sequnece = [2]
    if signal != None:
        sequnece.append(signal)

    dec_inp = build_input(sequnece)

    candidates = []
    options = []

    feed_dict = {enc_inputs: enc_inp, dec_inputs: dec_inp}
    values, indexs, state = sess.run([top_values, top_indexs, dec_states], feed_dict)

    for i in range(len(values)):
        candidates.append([values[i], [indexs[i]]])

    best_sequence = None
    highest_score = -sys.maxint - 1

    while True:

        # print candidates
        for i in range(len(candidates)):

            sequence = candidates[i][1]
            score = candidates[i][0]

            # if sequence end, evaluate
            if sequence[-1] == 3 or len(sequence) >= max_sequence_len:
                if score > highest_score:
                    highest_score = score
                    best_sequence = sequence
                continue

            # if not, continue searching
            dec_inp = build_input(sequence)

            feed_dict = {enc_states: state, dec_inputs: dec_inp}
            values, indexs = sess.run([top_values, top_indexs], feed_dict)

            for j in range(len(values)):
                new_sequence = list(sequence)
                new_sequence.append(indexs[j])
                options.append([score + values[j], new_sequence])

        # sort all options and keep top k
        options.sort(reverse=True)
        candidates = []

        for i in range(min(len(options), top_k)):
            if options[i][0] > highest_score:
                candidates.append(options[i])

        options = []
        if len(candidates) == 0:
            break

    if signal:
        best_sequence = [signal] + best_sequence

    return best_sequence[:-1]


def translate(token_list):
    enc = []
    for token in token_list:
        if token in reader.dict:
            enc.append(reader.dict[token])
        else:
            enc.append(reader.dict['[unk]'])
    # dec will be append with 2 inside the model
    return enc


# ---------------------------prediction loop---------------------------

if iteracitve:

    print("\n--------------------------")
    print("--Hello ErGo--")
    print("--------------------------\n")

    while True:
        try:
            line = sys.stdin.readline()
        except KeyboardInterrupt:
            print(
            "\nsession close")
            break

        token_list = re.findall(expression, line.lower())
        print("-------------")

        sequence = translate(token_list)
        enc_inp = build_input(sequence[::-1])
        response = beam_predict(enc_inp, None)
        sys.stdout.write('master: ')
        print_sentence(sequence)
        sys.stdout.write('-->: ')
        print_sentence(response)
        print(' ')

else:

    while True:
        try:
            data, index = reader.next_batch()
            enc_inp, dec_inp, dec_tar = s2s_reader.data_processing(data, buckets[index], batch_size)
            response = beam_predict(enc_inp)

            sys.stdout.write('src: ')
            print_sentence(data[0][0])
            sys.stdout.write('tar: ')
            print_sentence(data[0][1])
            sys.stdout.write('-->: ')
            print_sentence(response)
            print(' ')
        except KeyboardInterrupt:
            print(
            "\nsession close")
            break

sess.close()


inference/ergo.py
#!/usr/bin/env python3
# coding=utf-8
# Version:0.1
# Author: ysicing
# file: ergo.py
# time: 2017/7/18 23:24
# Copyright 2017 ysicing. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------

import collections
import os
import sys
import re
import numpy as np
import tensorflow as tf
#from models.model import rnn_model
#from dataset.chat import process_caht,generate_batch



def main(is_train):
    if is_train:
        print('[INFO] train chat...')
        from . import train
        print('[INFO] train end...')
    else:
        print('[INFO] chat with it...')
        from . import chat




inference/train.py
#!/usr/bin/env python3
# coding=utf-8
# Version:0.1
# Author: ysicing
# file: train.py
# time: 2017/7/19 21:38
# Copyright 2017 ysicing. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------

import tensorflow as tf
import json
import numpy as np
import os
import utils.s2s_reader as s2s_reader

# -----------------------------------parameters------------------------------------------

# the file where we read the data/model from
file_name = "dataset"
# if set to true, then load model from file instead of start a new model
load_model = False
# if set to true, use adam optimizer instead of sgd
adam_opt = True
# batch size for training
batch_size = 128

# data params
# bucket_option = [i for i in range(1, 20+1)]
bucket_option = [5, 10, 15, 20, 25, 31]
buckets = s2s_reader.create_bucket(bucket_option)

# load the data set into s2s_reader
# the total bucket numbers = bucket options number ^ 2
# if clean mode is true, the leftover data in the bucket will be used before the epoch is over
reader = s2s_reader.reader(file_name=file_name, batch_size=batch_size, buckets=buckets, bucket_option=bucket_option,
                           clean_mode=True)
vocab_size = len(reader.dict)

# if load_model = true, then we need to define the same parameter in the saved_model inorder to load it
hidden_size = 512
projection_size = 300
embedding_size = 300
num_layers = 1

# ouput_size for softmax layer
output_size = hidden_size
if projection_size != None:
    output_size = projection_size

# training params, truncated_norm will resample x > 2std; so when std = 0.1, the range of x is [-0.2, 0.2]
truncated_std = 0.1
keep_prob = 0.95
max_epoch = 20
norm_clip = 5

# training params for adam
adam_learning_rate = 0.001

# training params for sgd
learning_rate = 0.1
momentum = 0.9

# model name & save path
model_name = "p" + str(projection_size) + "_h" + str(hidden_size) + "_x" + str(num_layers)
save_path = file_name + "/" + model_name

# ---------------------------------model definition------------------------------------------

tf.reset_default_graph()
sess = tf.InteractiveSession()

# placeholder
enc_inputs = tf.placeholder(tf.int32, shape=(None, batch_size), name="enc_inputs")
targets = tf.placeholder(tf.int32, shape=(None, batch_size), name="targets")
dec_inputs = tf.placeholder(tf.int32, shape=(None, batch_size), name="dec_inputs")

# input embedding layers
emb_weights = tf.Variable(tf.truncated_normal([vocab_size, embedding_size], stddev=truncated_std), name="emb_weights")
enc_inputs_emb = tf.nn.embedding_lookup(emb_weights, enc_inputs, name="enc_inputs_emb")
dec_inputs_emb = tf.nn.embedding_lookup(emb_weights, dec_inputs, name="dec_inputs_emb")


# cell definiton

def getStackedLSTM():
    cell_list = []
    for i in range(num_layers):
        single_cell = tf.contrib.rnn.LSTMCell(
            num_units=hidden_size,
            num_proj=projection_size,
            # initializer=tf.truncated_normal_initializer(stddev=truncated_std),
            state_is_tuple=True
        )
        if i < num_layers - 1 or num_layers == 1:
            single_cell = tf.contrib.rnn.DropoutWrapper(cell=single_cell, output_keep_prob=keep_prob)
        cell_list.append(single_cell)
    return tf.contrib.rnn.MultiRNNCell(cells=cell_list, state_is_tuple=True)


# encoder & decoder defintion
_, enc_states = tf.nn.dynamic_rnn(cell=getStackedLSTM(),
                                  inputs=enc_inputs_emb,
                                  dtype=tf.float32,
                                  time_major=True,
                                  scope="encoder")

dec_outputs, dec_states = tf.nn.dynamic_rnn(cell=getStackedLSTM(),
                                            inputs=dec_inputs_emb,
                                            initial_state=enc_states,
                                            dtype=tf.float32,
                                            time_major=True,
                                            scope="decoder")

# output layers
project_w = tf.Variable(tf.truncated_normal(shape=[output_size, embedding_size], stddev=truncated_std),
                        name="project_w")
project_b = tf.Variable(tf.constant(shape=[embedding_size], value=0.1), name="project_b")
softmax_w = tf.Variable(tf.truncated_normal(shape=[embedding_size, vocab_size], stddev=truncated_std), name="softmax_w")
softmax_b = tf.Variable(tf.constant(shape=[vocab_size], value=0.1), name="softmax_b")

dec_outputs = tf.reshape(dec_outputs, [-1, output_size], name="dec_ouputs")
dec_proj = tf.matmul(dec_outputs, project_w) + project_b
logits = tf.nn.log_softmax(tf.matmul(dec_proj, softmax_w) + softmax_b, name="logits")

# loss function
flat_targets = tf.reshape(targets, [-1])
total_loss = tf.nn.sparse_softmax_cross_entropy_with_logits(logits=logits, labels=flat_targets)
avg_loss = tf.reduce_mean(total_loss)

# optimization
if adam_opt:
    optimizer = tf.train.AdamOptimizer(adam_learning_rate)
else:
    optimizer = tf.train.MomentumOptimizer(learning_rate=learning_rate, momentum=momentum, use_nesterov=True)

gvs = optimizer.compute_gradients(avg_loss)
capped_gvs = [(tf.clip_by_norm(grad, norm_clip), var) for grad, var in gvs]
train_op = optimizer.apply_gradients(capped_gvs)

# initialization or load model
saver = tf.train.Saver()

if load_model:
    cwd = os.getcwd()
    saver.restore(sess, cwd + "/" + save_path + "/model.ckpt")
    with open(save_path + '/summary.json') as json_data:
        losses = json.load(json_data)
        reader.epoch = len(losses) + 1
    print("Model restored.")
else:
    os.mkdir(save_path)
    sess.run(tf.global_variables_initializer())
    losses = []


# -----------------------------------training-------------------------------------------

def update_summary(save_path, losses):
    summary_location = save_path + "/summary.json"
    if os.path.exists(summary_location):
        os.remove(summary_location)
    with open(summary_location, 'w') as outfile:
        json.dump(losses, outfile)


# local variables
count = 0
epoch_loss = 0
epoch_count = 0

while True:

    curr_epoch = reader.epoch
    data, index = reader.next_batch()

    enc_inp, dec_inp, dec_tar = s2s_reader.data_processing(data, buckets[index], batch_size)

    if reader.epoch != curr_epoch:

        print(
        "\n----------end of epoch:" + str(reader.epoch - 1) + "----------" )
        print(
        "    avg loss: " + str(epoch_loss / epoch_count))
        print(
        "\n")

        losses.append(epoch_loss / epoch_count)

        epoch_loss = 0
        epoch_count = 0

        update_summary(save_path, losses)
        cwd = os.getcwd()
        saver.save(sess, cwd + "/" + save_path + "/model.ckpt")
        print("Model saved")

        if reader.epoch == (max_epoch + 1):
            break

    feed_dict = {enc_inputs: enc_inp, dec_inputs: dec_inp, targets: dec_tar}
    _, loss_t = sess.run([train_op, avg_loss], feed_dict)
    epoch_loss += loss_t

    count += 1
    epoch_count += 1

    if count % 10 == 0:
        print(
        str(loss_t) + " @ epoch: " + str(reader.epoch) + " count: " + str(epoch_count * batch_size))

sess.close()


utils/s2s_reader.py
#!/usr/bin/env python3
# coding=utf-8
# Version:0.1
# Author: ysicing
# file: s2s_reader.py.py
# time: 2017/7/18 23:31
# Copyright 2017 ysicing. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------


import os
import json
import numpy as np


def data_processing(data, size, batch_size):

	enc_len = size[0]
	dec_len = size[1]

	enc_inp = np.zeros((enc_len, batch_size))
	dec_inp = np.zeros((dec_len, batch_size))
	dec_tar = np.zeros((dec_len, batch_size))

	# pair[0] = encoder sequence, pair[1] = target sequence
	for i in range(len(data)):
		pair = data[i]
		# copy data to np-array
		enc_inp[enc_len-len(pair[0]):enc_len,i] = pair[0][::-1]
		dec_inp[1:len(pair[1])+1,i] = pair[1]
		dec_tar[0:len(pair[1]),i] = pair[1]
		# add start end token
		dec_inp[0,i]=2
		dec_tar[len(pair[1]),i]=3

	return enc_inp, dec_inp, dec_tar

def create_bucket (options):
	buckets=[]
	for i in range(len(options)):
		for j in range(len(options)):
			buckets.append((options[i],options[j]+1))
	return buckets



class reader:

	def __init__(self, file_name, batch_size, buckets, bucket_option, signal=False, clean_mode=False):

		self.epoch = 1
		self.batch_size = batch_size
		self.file_name = file_name
		self.buckets = buckets
		self.bucket_option = bucket_option
		self.output = []
		self.bucket_list = []
		self.clean_stock = False
		self.clean_mode = clean_mode
		self.bucket_dict = self.build_bucket_dict()

		for i in range (len(buckets)):
			self.bucket_list.append([])

		self.file = open(file_name + "/text.txt", "r")
		with open(file_name + "/dict.json", "r") as f:
			self.dict = json.load(f)

		if signal:
			with open(file_name + "/signal.json", "r") as f:
				self.signal_dict = json.load(f)


		self.id_dict = {}
		for key in self.dict:
			self.id_dict[self.dict[key]]=key



	#make batch for the model
	def next_batch(self):

		# normal mode
		if not self.clean_stock:
			index = self.fill_bucket()
			if index >= 0:
				output = self.bucket_list[index]
				#clean the bucket
				self.bucket_list[index] = []
				return output, index
			else:
				self.clean_stock = True
				for i in range(len(self.bucket_list)):
					if len(self.bucket_list[i]) > 0:
						output = self.bucket_list[i]
						self.bucket_list[i] = []
						return output, i

		# clean the stock
		elif self.clean_mode:
			for i in range(len(self.bucket_list)):
				if len(self.bucket_list[i]) > 0:
					output = self.bucket_list[i]
					self.bucket_list[i] = []
					return output, i
			#if all bucket are cleaned, reset the batch
			self.clean_stock = False
			self.reset()
			return self.next_batch()
		# ignore the stock
		else:
			for i in range(len(self.bucket_list)):
				self.bucket_list[i] = []
			self.clean_stock = False
			self.reset()
			return self.next_batch()

	#reset when epoch finished
	def reset(self):
		self.epoch += 1
		self.file.close()
		self.file = open(self.file_name + "/text.txt", "r")

	def build_bucket_dict(self):
		bucket_dict={}
		for i in range (1, self.bucket_option[-1]+1):
			count = len(self.bucket_option)-1
			for options in reversed(self.bucket_option):
				if options >= i:
					bucket_dict[i]=count
				count = count - 1
		return bucket_dict

	#pick bucket
	def check_bucket(self, pair):
		best_i = self.bucket_dict[len(pair[0])]
		best_j = self.bucket_dict[len(pair[1])]
		return best_i * len(self.bucket_option) + best_j


	#fill bucket
	def fill_bucket(self):
		while True:
			line = self.file.readline()
			if not line:
				break
			pair = json.loads(line)
			index = self.check_bucket(pair)
			#if size exceed all buckets, continue
			if index == -1:
				continue

			self.bucket_list[index].append(pair)
			if len(self.bucket_list[index]) == self.batch_size:
				return index

		#if hit the end of epoch
		return -1




main.py
#!/usr/bin/env python3
# coding=utf-8
# Version:0.0.1
# Author: ysicing
# file: main.py
# time: 2017/7/18 23:09
# Copyright 2017 ysicing. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------

import argparse

def parse_args():
    parse = argparse.ArgumentParser(description='Intelligent robot based on TensorFlow.')

    help_ = 'you can set this value in terminal,but now only support chat.'
    parse.add_argument('-t','--target',default='chat',choices=['chat'],help=help_)

    help_ = 'choose to train or test.'
    parse.add_argument('--train',dest='train',action='store_true',help=help_)
    parse.add_argument('--no-train',dest='train',action='store_false',help=help_)
    parse.set_defaults(train=True)

    args_ = parse.parse_args()
    return args_

if __name__ == '__main__':
    args =parse_args()
    if args.target == 'chat':
        from inference import ergo
        if args.train:
            ergo.main(True)
        else:
            ergo.main(False)
    elif args.target == 'poem':
        print('[INFO] The next update may be supported')
        pass
    else:
        print('[INFO] Simple english communion right now.')
